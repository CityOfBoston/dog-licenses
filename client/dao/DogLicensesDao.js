// @flow

import type { LoopbackGraphql } from '../loopback-graphql';
import type { DogLicense } from '../types';

import searchDogLicenses from '../queries/search-dog-licenses';
import type { DogLicenseSearchResults } from '../types';

export type DogLicenseCache = { [id: string]: DogLicense };

export default class DogLicensesDao {
  loopbackGraphql: LoopbackGraphql;

  constructor(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;
  }
  async search(
    firstName: string,
    lastName: string,
    dogName: string,
    year: number,
  ): Promise<DogLicenseSearchResults> {
    const results = await searchDogLicenses(
      this.loopbackGraphql,
      firstName,
      lastName,
      dogName,
      year,
    );

    return results;
  }

  parseQuery(
    fullQuery: string,
  ): { query: string, startYear: ?string, endYear: ?string } {
    // match a 4-digit year, and optionally a second 4-digit year with a hypen or en-dash
    const yearRegexp = /(\d{4})\s*[-â€“]?\s*(\d{4})?/;

    let query;
    let startYear;
    let endYear;

    const match = fullQuery.match(yearRegexp);
    if (match) {
      query = fullQuery.replace(yearRegexp, '').trim();
      startYear = match[1];
      endYear = match[2] || match[1];
    } else {
      query = fullQuery;
      startYear = null;
      endYear = null;
    }

    return {
      query,
      startYear,
      endYear,
    };
  }
}
