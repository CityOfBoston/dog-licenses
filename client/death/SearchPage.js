// @flow

import React from 'react';
import Head from 'next/head';
import Link from 'next/link';
import Router from 'next/router';
import type { Context } from 'next';

import type { ClientDependencies } from '../page';
import type { DeathCertificate } from '../types';

import type Cart from '../store/Cart';
import Nav from '../common/Nav';
import Pagination from '../common/Pagination';
import type { DeathCertificateSearchResults } from '../types';

import { GRAY_100 } from '../common/style-constants';

export type InitialProps = {|
  query: string,
  results: ?DeathCertificateSearchResults,
|}

export type Props = {
  /* :: ...InitialProps, */
  cart: Cart,
}

type State = {
  query: string,
};

export default class IndexPage extends React.Component {
  props: Props;
  state: State;

  static async getInitialProps(ctx: Context<*>, { deathCertificatesDao }: ClientDependencies): Promise<InitialProps> {
    const { query } = ctx;

    let results = null;

    if (query.q) {
      results = await deathCertificatesDao.search(query.q, parseInt(query.page, 10) || 1);
    }

    return {
      query: query.q || '',
      results,
    };
  }

  constructor(props: Props) {
    super(props);

    const { query } = props;

    this.state = {
      query,
    };
  }

  handleQueryChange = (ev: SyntheticInputEvent) => {
    this.setState({ query: ev.target.value });
  };

  handleSubmit = (ev: SyntheticInputEvent) => {
    const { query } = this.state;

    ev.preventDefault();
    Router.push(`/death?q=${encodeURIComponent(query)}`);
  }

  render() {
    const { results, cart } = this.props;
    const { query } = this.state;

    return (
      <div>
        <Head>
          <title>Boston.gov — Death Certificates</title>
        </Head>

        <Nav cart={cart} link="checkout" />

        <div className="p-a300">
          <div className="sh sh--b0">
            <h1 className="sh-title">Death Certificates</h1>
          </div>

          <form className="sf sf--md" acceptCharset="UTF-8" method="get" action="/death" onSubmit={this.handleSubmit}>
            <input name="utf8" type="hidden" value="✓" />

            <div className="sf-i">
              <input type="text" name="q" id="q" value={query} onChange={this.handleQueryChange} placeholder="Search…" className="sf-i-f" autoComplete="off" />
              <button className="sf-i-b" type="submit">Search</button>
            </div>
          </form>
        </div>

        { results && this.renderResults(results) }
        <div />

      </div>
    );
  }

  renderResults(results: DeathCertificateSearchResults) {
    // we want the query that was searched for
    const { query } = this.props;

    return (
      <div>
        <div className="p-a300 b--w">
          <div className="t--sans tt-u" style={{ fontSize: 12 }}>
            Showing { results.results.length } of { results.resultCount.toLocaleString() } results for “{query}”
          </div>
        </div>

        { results.results.map(this.renderResult) }

        { results.resultCount > results.results.length && this.renderPagination(results) }

        <div className="p-a300">
          Not finding what you’re looking for? Try refining your search or <a href="https://www.boston.gov/departments/registry/how-get-death-certificate" style={{ fontStyle: 'italic' }}>request a death certificate</a>.
        </div>
      </div>
    );
  }

  renderResult({ firstName, lastName, age, deathYear, id, causeOfDeath }: DeathCertificate) {
    return (
      <Link key={id} href={`/death/certificate?id=${id}`} as={`/death/certificate/${id}`}>
        <a className={'p-a300 br br-t100 b--w result'}>
          <div style={{ fontWeight: 'bold' }}>{firstName} {lastName}</div>
          <div>Died: {deathYear} Age: {age}</div>
          <div>Cause of death: {causeOfDeath ? 'Known' : 'Unknown'}</div>
          <style jsx>{`
            .result {
              display: block;
              color: inherit;
              font-style: italic;
              border-color: ${GRAY_100};
            }
          `}</style>
        </a>
      </Link>
    );
  }

  renderPagination({ page, pageCount }: DeathCertificateSearchResults) {
    const { query } = this.props;
    const makeHref = (p: number) => `/death?q=${query}&page=${p}`;

    return <Pagination page={page} pageCount={pageCount} hrefFunc={makeHref} />;
  }
}
