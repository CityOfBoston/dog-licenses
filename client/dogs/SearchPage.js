// @flow

import React from 'react';
import Head from 'next/head';
import type { Context } from 'next';

import type { ClientDependencies } from '../page';

import type { DogLicenseSearchResults } from '../types';

import SearchResult from './search/SearchResult';

export type InitialProps = {|
  results: ?DogLicenseSearchResults,
|};

export type Props = {
  ...InitialProps,
};

type State = {
  firstName: string,
  lastName: string,
  dogName: string,
  year: string,
};

export default class IndexPage extends React.Component {
  props: Props;
  state: State;

  static async getInitialProps(
    ctx: Context<*>,
    { dogLicensesDao }: ClientDependencies,
  ): Promise<InitialProps> {
    const { query } = ctx;

    let results = null;

    //THIS SEARCH FUNCTION IS NOT BEING USED NOW
    if (query.firstName && query.lastName && query.dogName && query.year) {
      results = await dogLicensesDao.search(
        query.firstName,
        query.lastName,
        query.dogName,
        parseInt(query.year, 10),
      );
      console.log('SearchPage: ', results);
    }

    return {
      results,
    };
  }

  constructor(props: Props) {
    super(props);
    //const { results } = props;

    this.state = {
      firstName: '',
      lastName: '',
      dogName: '',
      year: '',
    };
  }

  handleInputChange = (ev: SyntheticInputEvent) => {
    const target = ev.target;
    const value = target.value;
    const name = target.name;
    this.setState({ [name]: value });
  };

  render() {
    const { results } = this.props;

    return (
      <div>
        <Head>
          <title>Boston.gov — Dog Licenses</title>
        </Head>

        <div className="p-a300">
          <div className="sh sh--b0">
            <h1 className="sh-title">Renew Your Dog License</h1>
          </div>

          <form
            className="sf sf--md"
            acceptCharset="UTF-8"
            method="get"
            action="/dogs/form">
            <input name="utf8" type="hidden" value="✓" />

            <div className="txt">
              <label htmlFor="text" className="txt-l">Owner First Name</label>
              <input
                id="text"
                name="firstName"
                type="text"
                value={this.state.firstName}
                onChange={this.handleInputChange}
                placeholder="Emily"
                className="txt-f"
              />
            </div>

            <div className="txt">
              <label htmlFor="text" className="txt-l">Owner Last Name</label>
              <input
                id="text"
                type="text"
                name="lastName"
                value={this.state.lastName}
                onChange={this.handleInputChange}
                placeholder="Elizabeth"
                className="txt-f"
              />
            </div>

            <div className="txt">
              <label htmlFor="text" className="txt-l">Dog name</label>
              <input
                id="text"
                type="text"
                name="dogName"
                value={this.state.dogName}
                onChange={this.handleInputChange}
                placeholder="Clifford"
                className="txt-f"
              />
            </div>

            <div className="txt">
              <label htmlFor="text" className="txt-l">Year licensed</label>
              <input
                id="text"
                type="text"
                name="year"
                value={this.state.year}
                onChange={this.handleInputChange}
                placeholder="2016"
                className="txt-f"
              />
            </div>

            <button className="btn" type="submit">Search for id</button>

          </form>
        </div>

        {results && this.renderResults(results)}
        <div />
      </div>
    );
  }

  renderResults(results: DogLicenseSearchResults) {
    // we want the query that was searched for
    // const start = 1 + (results.page - 1) * results.pageSize;
    // const end = Math.min(start + results.pageSize - 1, results.resultCount);

    return (
      <div>

        <div className="p-a300 b--w">
          <div className="t--sans tt-u" style={{ fontSize: 12 }}>
            Showing results:
          </div>
        </div>

        {results.map(license =>
          <SearchResult license={license} key={license.id} />,
        )}

        <div className="p-a300">
          Not finding what you’re looking for? Try refining your search or{' '}
          <a
            href="https://www.boston.gov/departments/registry/how-get-dog-license"
            style={{ fontStyle: 'italic' }}>
            request a dog license
          </a>.
        </div>

        <form
          className="sf sf--md"
          acceptCharset="UTF-8"
          method="get"
          action="/dogs/form">
          <input name="utf8" type="hidden" value="✓" />

          <div className="m-v400 m-h200">
            <button className="btn" type="submit">Renew This license</button>
          </div>

        </form>
      </div>
    );
  }
}
