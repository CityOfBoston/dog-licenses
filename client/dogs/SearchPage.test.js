// @flow
import React from 'react';
import renderer from 'react-test-renderer';

//import type { DogLicense, DogLicenseSearchResults } from '../types';
import DogLicensesDao from '../dao/DogLicensesDao';

import SearchPage from './SearchPage';
import type { InitialProps } from './SearchPage';
import // TYPICAL_LICENSE,
// PENDING_LICENSE,
// NO_DATE_LICENSE,
'../../fixtures/client/dog-licenses';

jest.mock('next/router');
jest.mock('../dao/DogLicensesDao');

// const TEST_DOG_LICENSES: DogLicense[] = [
//   TYPICAL_LICENSE,
//   PENDING_LICENSE,
//   NO_DATE_LICENSE,
// ];

// const TEST_SEARCH_RESULTS: DogLicenseSearchResults = {
//   id: 'A08392',
//   firstName: 'ALEX',
//   lastName: 'ABOUD',
//   dogName: 'LUCY',
//   year: 2016,
// };

const renderFromInitialProps = async (
  query: { [key: string]: string },
  dependencies: Object,
) => {
  // const cart = new Cart();

  const initialProps: InitialProps = await SearchPage.getInitialProps(
    ({
      query,
    }: any),
    dependencies,
  );

  return renderer.create(<SearchPage /*cart={cart} */ {...initialProps} />);
};

describe('rendering', () => {
  let dogLicensesDao;

  beforeEach(() => {
    dogLicensesDao = new DogLicensesDao((null: any));
  });

  it('shows empty search box', async () => {
    expect(
      (await renderFromInitialProps({}, { dogLicensesDao })).toJSON(),
    ).toMatchSnapshot();
  });

  // it('shows search results', async () => {
  //   dogLicensesDao.search.mockReturnValue(TEST_SEARCH_RESULTS);

  //   expect(
  //     (await renderFromInitialProps(
  //       { q: 'Monkey Joe' },
  //       { dogLicensesDao },
  //     )).toJSON(),
  //   ).toMatchSnapshot();
  //   expect(dogLicensesDao.search).toHaveBeenCalledWith('Monkey Joe', 1);
  // });
});
