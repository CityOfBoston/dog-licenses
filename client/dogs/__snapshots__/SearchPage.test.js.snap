// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering shows empty search box 1`] = `
<div>
  <div
    className="p-a300"
  >
    <div
      className="sh sh--b0"
    >
      <h1
        className="sh-title"
      >
        Renew Your Dog License
      </h1>
    </div>
    <form
      acceptCharset="UTF-8"
      action="/dogs/form"
      className="sf sf--md"
      method="get"
    >
      <input
        name="utf8"
        type="hidden"
        value="âœ“"
      />
      <div
        className="txt"
      >
        <label
          className="txt-l"
          htmlFor="text"
        >
          Owner First Name
        </label>
        <input
          className="txt-f"
          id="text"
          name="firstName"
          onChange={[Function]}
          placeholder="Emily"
          type="text"
          value=""
        />
      </div>
      <div
        className="txt"
      >
        <label
          className="txt-l"
          htmlFor="text"
        >
          Owner Last Name
        </label>
        <input
          className="txt-f"
          id="text"
          name="lastName"
          onChange={[Function]}
          placeholder="Elizabeth"
          type="text"
          value=""
        />
      </div>
      <div
        className="txt"
      >
        <label
          className="txt-l"
          htmlFor="text"
        >
          Dog name
        </label>
        <input
          className="txt-f"
          id="text"
          name="dogName"
          onChange={[Function]}
          placeholder="Clifford"
          type="text"
          value=""
        />
      </div>
      <div
        className="txt"
      >
        <label
          className="txt-l"
          htmlFor="text"
        >
          Year licensed
        </label>
        <input
          className="txt-f"
          id="text"
          name="year"
          onChange={[Function]}
          placeholder="2016"
          type="text"
          value=""
        />
      </div>
      <button
        className="btn"
        type="submit"
      >
        Search for id
      </button>
    </form>
  </div>
  <div />
</div>
`;
